File: backend/app/schemas/community.py
"""Pydantic schemas for the peer Doubt Community (Phase 17, F4)."""
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class DoubtCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    body: str = Field(..., min_length=1, max_length=5000)
    tags: list[str] = Field(default_factory=list)
    subject_id: uuid.UUID | None = None


class DoubtAnswerCreate(BaseModel):
    answer_text: str = Field(..., min_length=1, max_length=5000)


class DoubtAnswerResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    doubt_id: uuid.UUID
    answered_by_id: uuid.UUID | None
    answered_by_name: str | None = None
    answer_text: str
    is_ai_generated: bool
    is_accepted: bool
    upvote_count: int
    created_at: datetime


class DoubtResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    student_id: uuid.UUID
    student_name: str | None = None
    subject_id: uuid.UUID | None = None
    subject_code: str | None = None
    title: str
    body: str
    tags: list[str]
    is_resolved: bool
    upvote_count: int
    view_count: int
    answer_count: int
    has_ai_answer: bool
    created_at: datetime


class DoubtDetailResponse(DoubtResponse):
    answers: list[DoubtAnswerResponse]

File: frontend/src/types/community.ts
/**
 * Peer Doubt Community types — mirror backend Pydantic schemas in
 * backend/app/schemas/community.py
 */

export interface DoubtAnswerResponse {
  id: string
  doubt_id: string
  answered_by_id: string | null
  answered_by_name: string | null
  answer_text: string
  is_ai_generated: boolean
  is_accepted: boolean
  upvote_count: number
  created_at: string
}

export interface DoubtResponse {
  id: string
  student_id: string
  student_name: string | null
  subject_id: string | null
  subject_code: string | null
  title: string
  body: string
  tags: string[]
  is_resolved: boolean
  upvote_count: number
  view_count: number
  answer_count: number
  has_ai_answer: boolean
  created_at: string
}

export interface DoubtDetailResponse extends DoubtResponse {
  answers: DoubtAnswerResponse[]
}

export interface DoubtCreate {
  title: string
  body: string
  tags?: string[]
  subject_id?: string | null
}

export interface DoubtAnswerCreate {
  answer_text: string
}
